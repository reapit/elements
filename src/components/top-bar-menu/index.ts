export * from './top-bar-menu'
export * from './styles'
export * from './top-bar-menu.atoms'
export * from './top-bar-menu-context'
