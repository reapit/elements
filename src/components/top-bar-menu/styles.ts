import { styled } from '@linaria/react'
import { isMobile } from '#src/styles/media'

export const ElTopBarMenuHeader = styled.div`
  display: flex;
  padding: var(--spacing-2) var(--spacing-4);
  justify-content: flex-end;
  align-items: center;
  align-self: stretch;
  border-bottom: var(--border-width-default) solid var(--outline-default);
  background: var(--comp-navigation-colour-fill-top_bar);
`

export const ElTopBarMenuBody = styled.nav`
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  align-self: stretch;
  scrollbar-gutter: stable;
  overflow-y: auto;
  padding: var(--spacing-3) var(--spacing-none);
  background: var(--comp-navigation-colour-fill-mobile_nav-default);
`

export const ElTopBarMenuList = styled.ul`
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  align-self: stretch;
  padding: var(--spacing-3);
  gap: var(--spacing-1);
`

export const ElTopBarMenu = styled.dialog`
  // NOTE: overrides default browser styles for 'dialog'
  max-width: 100%;
  max-height: 100%;
  width: 375px;
  height: 100%;
  z-index: 99;
  display: none;
  border: none;
  padding: var(--spacing-none);
  animation: topBarMenuSlideRight 0.2s ease-out;
  ${isMobile} {
    width: 100%;
    margin: var(--spacing-none);
    scrollbar-gutter: unset;
    animation: topBarMenuFadeOut 0.2s ease-out;
  }
  &[open] {
    animation: topBarMenuSlideLeft 0.2s ease-out forwards;
    display: flex;
    flex-direction: column;
    ${isMobile} {
      animation: topBarMenuFadeIn 0.2s ease-out forwards;
    }
    &::backdrop {
      background-color: var(--black);
      opacity: 0.5;
      animation: backdropFadeIn 0.2s ease-out;
    }
  }
  ${ElTopBarMenuList}:not(:first-child) {
    border-top: var(--border-width-default) solid var(--outline-default);
  }
  @keyframes topBarMenuFadeIn {
    from {
      opacity: 0;
      display: none;
    }
    to {
      opacity: 1;
      display: flex;
    }
  }
  @keyframes topBarMenuFadeOut {
    from {
      opacity: 1;
      display: flex;
    }
    to {
      opacity: 0;
      display: none;
    }
  }
  @keyframes topBarMenuSlideLeft {
    from {
      margin-right: -375px;
      display: none;
    }
    to {
      margin-right: 0px;
      display: flex;
    }
  }
  @keyframes topBarMenuSlideRight {
    from {
      margin-right: 0px;
      display: flex;
    }
    to {
      margin-right: -375px;
      display: none;
    }
  }
  @keyframes backdropFadeIn {
    from {
      background-color: var(--black);
      opacity: 0;
    }
    to {
      background-color: var(--black);
      opacity: 0.5;
    }
  }
`
