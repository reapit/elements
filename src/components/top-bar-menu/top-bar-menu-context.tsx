import { createContext, ReactNode, useContext, type FC } from 'react'
import { TopBarMenuProps } from './top-bar-menu'

type TopBarMenuContextModel = Pick<TopBarMenuProps, 'onClose'>

interface TopBarMenuProviderProps extends TopBarMenuContextModel {
  children: ReactNode
}

const TopBarMenuContext = createContext<TopBarMenuContextModel | null>(null)

const TopBarMenuProvider: FC<TopBarMenuProviderProps> = ({ children, onClose }) => {
  return <TopBarMenuContext.Provider value={{ onClose }}>{children}</TopBarMenuContext.Provider>
}

export { TopBarMenuProvider }

export const useTopBarMenuContext = () => {
  const context = useContext(TopBarMenuContext)
  if (!context) {
    throw new Error('useTopBarMenuContext must be used within a TopBarMenuProvider')
  }
  return context
}
