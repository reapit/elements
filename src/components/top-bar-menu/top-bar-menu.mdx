import { Meta, Canvas, ArgTypes, Description } from '@storybook/blocks'
import { RenderHtmlMarkup } from '../../storybook/render-html-markup'
import * as TopBarMenu from './top-bar-menu.stories'

<Meta of={TopBarMenu} />

# Top Bar Menu

The Top Bar Menu component is a mobile-friendly navigation panel that slides in from the right side of the screen. It provides a structured way to organize navigation items in lists and collapsible groups.

## Component Structure

The TopBarMenu uses a compound component pattern, exposing these subcomponents:

- `TopBarMenu.Header` - Header section with a close button
- `TopBarMenu.Body` - Container for the menu content
- `TopBarMenu.List` - Groups of menu items, separated by borders
- `TopBarMenu.Group` - Collapsible groups of menu items with a label
- `TopBarMenu.Item` - Individual navigation items (links or buttons)

## Default Example

The basic implementation of the TopBarMenu with main navigation, settings, and user menu sections.

<Canvas of={TopBarMenu.Default} />

## Initially Open

Shows how the TopBarMenu looks when initially opened.

<Canvas of={TopBarMenu.InitiallyOpen} />

## With Active Item

Demonstrates how active items are highlighted in the menu.

<Canvas of={TopBarMenu.WithActiveItem} />

## With Notification Badges

Shows how notification badges appear on menu items and groups.

<Canvas of={TopBarMenu.WithNotificationBadges} />

## Mobile View

How the TopBarMenu appears on mobile devices.

<Canvas of={TopBarMenu.MobileView} />

## Usage Example

```tsx
import { TopBarMenu } from '@reapit/elements'
import { useState } from 'react'

const MyNavigation = () => {
  const [isOpen, setIsOpen] = useState(false)

  const handleClose = () => setIsOpen(false)

  return (
    <>
      <button onClick={() => setIsOpen(true)}>Open Menu</button>

      <TopBarMenu isOpen={isOpen} onClose={handleClose}>
        <TopBarMenu.Header />
        <TopBarMenu.Body>
          <TopBarMenu.List>
            <TopBarMenu.Item href="/dashboard" isActive>
              Dashboard
            </TopBarMenu.Item>

            <TopBarMenu.Group label="Properties" hasBadge>
              <TopBarMenu.Item href="/properties">All Properties</TopBarMenu.Item>
              <TopBarMenu.Item href="/property-search">Property Search</TopBarMenu.Item>
              <TopBarMenu.Item href="/saved-properties" hasBadge>
                Saved Properties
              </TopBarMenu.Item>
            </TopBarMenu.Group>

            <TopBarMenu.Item href="/calendar">Calendar</TopBarMenu.Item>
          </TopBarMenu.List>

          <TopBarMenu.List>
            <TopBarMenu.Item href="/profile">My Profile</TopBarMenu.Item>
            <TopBarMenu.Item onClick={handleLogout}>Sign Out</TopBarMenu.Item>
          </TopBarMenu.List>
        </TopBarMenu.Body>
      </TopBarMenu>
    </>
  )
}
```

## Props

<ArgTypes of={TopBarMenu} />
