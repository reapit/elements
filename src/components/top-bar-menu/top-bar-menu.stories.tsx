import { useState } from 'react'

import type { Meta, StoryObj } from '@storybook/react'
import { Button } from '../button'

import { TopBarMenu, TopBarMenuProps } from './top-bar-menu'

const meta = {
  title: 'Components/Top Bar Menu',
  component: TopBarMenu,
  parameters: {
    layout: 'padded',
  },
  args: {
    isOpen: true,
    children: null,
  },
  render: ({}) => {
    const [isOpen, setIsOpen] = useState(false)

    const handleOnClose = () => {
      setIsOpen(false)
    }

    return (
      <>
        <Button onClick={() => setIsOpen(true)}>{isOpen ? 'Close' : 'Open'}</Button>
        <TopBarMenu isOpen={isOpen} onClose={handleOnClose}>
          <TopBarMenu.Header />
          <TopBarMenu.Body>
            <TopBarMenu.List>
              <TopBarMenu.Group label="Main Nav Item 1" hasBadge>
                <TopBarMenu.Item href="#item-1.1">Label</TopBarMenu.Item>
                <TopBarMenu.Item href="#item-1.2">Label</TopBarMenu.Item>
                <TopBarMenu.Item href="#item-1.3">Label</TopBarMenu.Item>
                <TopBarMenu.Item href="#item-1.4">Label</TopBarMenu.Item>
                <TopBarMenu.Item href="#item-1.5">Label</TopBarMenu.Item>
              </TopBarMenu.Group>
              <TopBarMenu.Item href="#item-2">Main Nav Item 2</TopBarMenu.Item>
              <TopBarMenu.Item href="#item-3">Main Nav Item 3</TopBarMenu.Item>
              <TopBarMenu.Item href="#item-4">Main Nav Item 4</TopBarMenu.Item>
              <TopBarMenu.Item href="#item-5">Main Nav Item 5</TopBarMenu.Item>
            </TopBarMenu.List>
            <TopBarMenu.List>
              <TopBarMenu.Group label="Secondary Nav Item 1" hasBadge isActive>
                <TopBarMenu.Item href="#item-1.1">Label</TopBarMenu.Item>
                <TopBarMenu.Item href="#item-1.2">Label</TopBarMenu.Item>
                <TopBarMenu.Item href="#item-1.3" isActive>
                  Label
                </TopBarMenu.Item>
                <TopBarMenu.Item href="#item-1.4" hasBadge>
                  Label
                </TopBarMenu.Item>
                <TopBarMenu.Item href="#item-1.5">Label</TopBarMenu.Item>
              </TopBarMenu.Group>
              <TopBarMenu.Item href="#item-2">Secondary Nav Item 2</TopBarMenu.Item>
              <TopBarMenu.Item href="#item-3">Secondary Nav Item 3</TopBarMenu.Item>
              <TopBarMenu.Item href="#item-4">Secondary Nav Item 4</TopBarMenu.Item>
              <TopBarMenu.Item href="#item-5">Secondary Nav Item 5</TopBarMenu.Item>
            </TopBarMenu.List>
            <TopBarMenu.List>
              <TopBarMenu.Item href="#item-1.1">User Item 1</TopBarMenu.Item>
              <TopBarMenu.Item href="#item-1.1">User Item 2</TopBarMenu.Item>
              <TopBarMenu.Item href="#item-1.1">User Item 3</TopBarMenu.Item>
              <TopBarMenu.Item href="#item-1.1">User Item 4</TopBarMenu.Item>
              <TopBarMenu.Item href="#item-1.1">User Item 5</TopBarMenu.Item>
            </TopBarMenu.List>
          </TopBarMenu.Body>
        </TopBarMenu>
      </>
    )
  },
} satisfies Meta<TopBarMenuProps>

export default meta

type Story = StoryObj<typeof meta>

export const Desktop: Story = {
  parameters: {
    viewport: {
      defaultViewport: 'desktop',
    },
  },
}

export const Mobile: Story = {
  parameters: {
    viewport: {
      defaultViewport: 'mobile2',
    },
  },
}
