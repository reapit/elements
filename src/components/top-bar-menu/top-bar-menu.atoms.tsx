import { FC } from 'react'
import { NavIconItem } from '../nav-icon-item'
import { Icon } from '../icon'
import { ElTopBarMenuHeader } from './styles'
import { useTopBarMenuContext } from './top-bar-menu-context'

export const TopBarMenuHeader: FC = () => {
  const { onClose } = useTopBarMenuContext()
  return (
    <ElTopBarMenuHeader>
      <NavIconItem aria-label="Close" icon={<Icon icon="close" fontSize="24px" />} onClick={onClose} />
    </ElTopBarMenuHeader>
  )
}
