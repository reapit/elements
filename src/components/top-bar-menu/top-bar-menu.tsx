import { FC, ReactNode, useRef } from 'react'

import { useToggleDialogVisibilityEffect } from '../dialog'
import { TopBarMenuItem } from '../top-bar-menu-item'
import { TopBarMenuItemGroup } from '../top-bar-menu-item-group'

import { TopBarMenuHeader } from './top-bar-menu.atoms'
import { ElTopBarMenu, ElTopBarMenuBody, ElTopBarMenuList } from './styles'
import { TopBarMenuProvider } from './top-bar-menu-context'

export interface TopBarMenuProps {
  isOpen: boolean
  children: ReactNode
  onClose?: VoidFunction
}

type TopBarMenuFC = FC<TopBarMenuProps> & {
  Header: typeof TopBarMenuHeader
  Body: typeof ElTopBarMenuBody
  List: typeof ElTopBarMenuList
  Group: typeof TopBarMenuItemGroup
  Item: typeof TopBarMenuItem
}

const TopBarMenu: TopBarMenuFC = ({ children, isOpen, onClose }) => {
  const ref = useRef<HTMLDialogElement>(null)

  useToggleDialogVisibilityEffect({ isOpen, ref, onClose })

  return (
    <TopBarMenuProvider onClose={onClose}>
      <ElTopBarMenu ref={ref}>{children}</ElTopBarMenu>
    </TopBarMenuProvider>
  )
}

TopBarMenu.Header = TopBarMenuHeader
TopBarMenu.Body = ElTopBarMenuBody
TopBarMenu.List = ElTopBarMenuList
TopBarMenu.Group = TopBarMenuItemGroup
TopBarMenu.Item = TopBarMenuItem

export { TopBarMenu }
