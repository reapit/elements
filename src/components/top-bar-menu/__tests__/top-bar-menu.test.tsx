import { fireEvent, render } from '@testing-library/react'
import { TopBarMenu } from '../top-bar-menu'

vi.mock('../../icon', () => ({
  Icon: () => <svg data-name="mocked" />,
}))

describe('TopBarMenu', () => {
  it('should match a snapshot when isOpen state is true', () => {
    expect(
      render(
        <TopBarMenu isOpen onClose={vi.fn()}>
          <TopBarMenu.Header />
          <TopBarMenu.Body>
            <TopBarMenu.List>
              <TopBarMenu.Group label="Main Nav Item 1" hasBadge>
                <TopBarMenu.Item href="#item-1.1">Label</TopBarMenu.Item>
                <TopBarMenu.Item href="#item-1.2">Label</TopBarMenu.Item>
              </TopBarMenu.Group>
              <TopBarMenu.Item href="#item-2">Main Nav Item 2</TopBarMenu.Item>
            </TopBarMenu.List>
            <TopBarMenu.List>
              <TopBarMenu.Item href="#item-1.1">User Item 1</TopBarMenu.Item>
              <TopBarMenu.Item href="#item-1.1">User Item 2</TopBarMenu.Item>
            </TopBarMenu.List>
          </TopBarMenu.Body>
        </TopBarMenu>,
      ).asFragment(),
    ).toMatchSnapshot()
  })

  it('should handle the keyboard navigation properly', () => {
    const mockOnButtonClick = vi.fn()
    const mockOnKeyDown = vi.fn()

    render(
      <TopBarMenu isOpen onKeyDown={mockOnKeyDown}>
        <TopBarMenu.Body>
          <TopBarMenu.List>
            <TopBarMenu.Group isActive={false} label="Main Nav Item 1">
              <TopBarMenu.Item href="#item-1.1">Sub Nav Item 1</TopBarMenu.Item>
              <TopBarMenu.Item href="#item-1.2">Sub Nav Item 2</TopBarMenu.Item>
            </TopBarMenu.Group>
            <TopBarMenu.Item onClick={mockOnButtonClick}>Main Nav Item 2</TopBarMenu.Item>
          </TopBarMenu.List>
        </TopBarMenu.Body>
      </TopBarMenu>,
    )

    const getItems = () =>
      document?.querySelectorAll(
        'ul:not([aria-hidden="true"]):not([aria-hidden="false"]) > li > a, ul[aria-hidden="false"] > li > a, button',
      ) as NodeListOf<HTMLElement>

    const initialItems = getItems()
    expect(initialItems).toHaveLength(2)

    initialItems[0].focus()

    // NOTE: proving the sub items are unaccessible while the parent item is collapsed
    fireEvent.keyDown(document.activeElement!, { key: 'ArrowDown' })
    expect(document.activeElement).toBe(initialItems[1])
    fireEvent.keyDown(document.activeElement!, { key: 'ArrowUp' })
    expect(document.activeElement).toBe(initialItems[0])
    fireEvent.keyDown(document.activeElement!, { key: 'ArrowRight' })
    expect(document.activeElement).toBe(initialItems[0])

    // NOTE: ensure the sub items are accessible while the parent item expanded
    const expandedItems = getItems()
    expect(expandedItems).toHaveLength(4)
    fireEvent.keyDown(document.activeElement!, { key: 'ArrowDown' })
    expect(document.activeElement).toBe(expandedItems[1])
    fireEvent.keyDown(document.activeElement!, { key: 'ArrowDown' })
    expect(document.activeElement).toBe(expandedItems[2])
    fireEvent.keyDown(document.activeElement!, { key: 'ArrowDown' })
    expect(document.activeElement).toBe(expandedItems[3])

    // NOTE: move back to the first item
    fireEvent.keyDown(document.activeElement!, { key: 'ArrowDown' })
    expect(document.activeElement).toBe(expandedItems[0])

    // NOTE: asserting the expand and collapse functionality with across key strokes
    fireEvent.keyDown(document.activeElement!, { key: 'ArrowLeft' })
    expect(getItems()).toHaveLength(2)
    fireEvent.keyDown(document.activeElement!, { key: 'Enter' })
    expect(getItems()).toHaveLength(4)
    fireEvent.keyDown(document.activeElement!, { key: ' ' })
    expect(getItems()).toHaveLength(2)
  })
})
