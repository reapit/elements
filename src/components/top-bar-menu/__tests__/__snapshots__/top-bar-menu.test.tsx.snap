// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TopbarMenu > should match a snapshot when isOpen state is true 1`] = `
<DocumentFragment>
  <dialog
    class="mocked-styled-19 el-top-bar-menu"
    open=""
  >
    <div
      class="mocked-styled-16 el-top-bar-menu-header"
    >
      <button
        aria-label="Close"
        class="mocked-styled-14 el-button-nav-icon-item"
      >
        <svg
          data-name="mocked"
        />
      </button>
    </div>
    <nav
      class="mocked-styled-17 el-top-bar-menu-body"
    >
      <ul
        class="mocked-styled-18 el-top-bar-menu-list"
      >
        <li
          class="mocked-styled-12 el-top-bar-menu-item-group-list-item"
          data-is-expanded="false"
        >
          <button
            aria-controls=":r0:"
            aria-expanded="false"
            aria-label="Main Nav Item 1"
            class="mocked-styled-9 el-top-bar-menu-item-button el-top-bar-menu-item-group-button el-top-bar-menu-item-button"
            type="button"
          >
            <div
              class="mocked-styled-10 el-top-bar-menu-item-group-content"
            >
              Main Nav Item 1
              <span
                class="mocked-styled-7 el-top-bar-menu-item-badge"
              />
            </div>
            <svg
              data-name="mocked"
            />
          </button>
          <ul
            aria-hidden="true"
            class="mocked-styled-11 el-top-bar-menu-item-group-list"
            id=":r0:"
          >
            <li
              class="mocked-styled-8 el-top-bar-menu-item-list-item"
            >
              <a
                class="mocked-styled-5 el-top-bar-menu-item-anchor"
                href="#item-1.1"
              >
                Label
              </a>
            </li>
            <li
              class="mocked-styled-8 el-top-bar-menu-item-list-item"
            >
              <a
                class="mocked-styled-5 el-top-bar-menu-item-anchor"
                href="#item-1.2"
              >
                Label
              </a>
            </li>
          </ul>
        </li>
        <li
          class="mocked-styled-8 el-top-bar-menu-item-list-item"
        >
          <a
            class="mocked-styled-5 el-top-bar-menu-item-anchor"
            href="#item-2"
          >
            Main Nav Item 2
          </a>
        </li>
      </ul>
      <ul
        class="mocked-styled-18 el-top-bar-menu-list"
      >
        <li
          class="mocked-styled-8 el-top-bar-menu-item-list-item"
        >
          <a
            class="mocked-styled-5 el-top-bar-menu-item-anchor"
            href="#item-1.1"
          >
            User Item 1
          </a>
        </li>
        <li
          class="mocked-styled-8 el-top-bar-menu-item-list-item"
        >
          <a
            class="mocked-styled-5 el-top-bar-menu-item-anchor"
            href="#item-1.1"
          >
            User Item 2
          </a>
        </li>
      </ul>
    </nav>
  </dialog>
</DocumentFragment>
`;
